core/service.go

```diff
@@ -28,117 +32,195 @@ import (
 
 var _ services.Service = &Service{}
 
-// Service is a overhead layer that allows for communication between the runtime, BABE, and the p2p layer.
-// It deals with the validation of transactions and blocks by calling their respective validation functions
-// in the runtime.
+// Service is an overhead layer that allows communication between the runtime,
+// BABE session, and p2p service. It deals with the validation of transactions
+// and blocks by calling their respective validation functions in the runtime.
 type Service struct {
-	rt *runtime.Runtime
-	b  *babe.Session
-
-	msgRec  <-chan p2p.Message
-	msgSend chan<- p2p.Message
+	rt      *runtime.Runtime
+	bs      *babe.Session
+	bsRec   <-chan types.Block // receive blocks from BABE session
+	p2pRec  <-chan p2p.Message // receive messages from p2p service
+	p2pSend chan<- p2p.Message // send messages to p2p service
 }
 
-// NewService returns a Service that connects the runtime, BABE, and the p2p messages.
-func NewService(rt *runtime.Runtime, msgRec <-chan p2p.Message, msgSend chan<- p2p.Message) (*Service, error) {
-	b, err := babe.NewSession([32]byte{}, [64]byte{}, rt, msgSend)
+// NewService returns a new core service that connects the runtime, BABE session, and p2p service.
+func NewService(rt *runtime.Runtime, bsRec chan types.Block, p2pRec <-chan p2p.Message, p2pSend chan<- p2p.Message) (*Service, error) {
+
+	// create a new BABE session (bsRec becomes blockSend in babe)
+	bs, err := babe.NewSession([32]byte{}, [64]byte{}, rt, bsRec)
 	if err != nil {
 		return nil, err
 	}
 
-	return &Service{
+	s := &Service{
 		rt:      rt,
-		b:       b,
-		msgRec:  msgRec,
-		msgSend: msgSend,
-	}, nil
+		bs:      bs,
+		bsRec:   bsRec,
+		p2pRec:  p2pRec,
+		p2pSend: p2pSend,
+	}
+
+	return s, nil
 }
 
-// Start begins the service. This begins watching the message channel for new block or transaction messages.
+// Start starts the core service
 func (s *Service) Start() error {
-	e := make(chan error)
-	go s.start(e)
-	return <-e
-}
 
-func (s *Service) start(e chan error) {
-	e <- nil
+	// start receiving blocks from BABE session
+	go s.startReceivingBlocks()
 
-	for {
-		msg, ok := <-s.msgRec
-		if !ok {
-			log.Warn("core service message watcher", "error", "channel closed")
-			break
-		}
+	// start receiving messages from p2p service
+	go s.startReceivingMessages()
```

Comment:
I think `start` is a bit of a misnomer. These functions **are** receiving blocks and messages, you start them by putting them in their own routine.

Reply:
Ah, ok. Maybe just `receiveMessages` then.